cpp/src/io/orc/stats_enc.cu

```diff
@@ -0,0 +1,380 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "orc_common.h"
+#include "orc_gpu.h"
+#include <io/utilities/block_utils.cuh>
+
+namespace cudf {
+namespace io {
+namespace orc {
+namespace gpu {
+
+/**
+ * @brief Initializes statistics groups
+ *
+ * @param[out] groups Statistics groups
+ * @param[in] cols Column descriptors
+ * @param[in] num_columns Number of columns
+ * @param[in] num_rowgroups Number of rowgroups
+ * @param[in] row_index_stride Rowgroup size in rows
+ *
+ **/
+// blockDim {128,1,1}
+__global__ void __launch_bounds__(128)
+gpuInitStatisticsGroups(statistics_group *groups, const stats_column_desc *cols,
+                        uint32_t num_columns, uint32_t num_rowgroups, uint32_t row_index_stride) {
+  __shared__ __align__(4) volatile statistics_group grp_g[4];
```

Comment:
Here you hard-code `4` statistics_groups per thread block. But below you decide which threads to use based on `sizeof(uint32_t)` rather than `4`. I suggest instead defining a constexpr variable with a clear name indicating groups per block.

Reply:
The number of thread is *not* decided based on sizeof(uint32_t)

Reply:
The point is that this code needs either documentation or clarification. A person other than you, or you a few months from now, will not know why this is hard-coded to 4.

Reply:
I supposed I could have used 128/32 to make it more obvious where the 4 comes from (both the 128 and 32 constants appears within these 3 consecutive lines of code without having to scroll to see various definitions). Now using a constexpr threads_per_block / threads_per_chunk.